### PR TITLE
typescript-node: Fix cookie auth, fix multiple API key auth

### PR DESCRIPTION
**Pull request by janjongboom**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

TC: @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @nicokoenig (2018/09) @topce  @akehir

### Description of the PR

This pull request fixes two issues with the typescript-node generator:

1. If multiple API Key authentication schemas are defined the generator emits code that won't compile.
1. `in: cookie` is not supported.

### Multiple API Key authentication schemas

If you have the following OpenAPI definition:

```
components:
    securitySchemes:
        ApiKeyAuthentication:
            type: apiKey
            in: header
            name: x-api-key
        JWTAuthentication:
            type: apiKey
            in: cookie
            name: jwt
```

The generator emits the following code:

```js
import { ApiKeyAuth } from '../model/models';		
import { ApiKeyAuth } from '../model/models';		
```

This won't compile. In `api-single.mustache` there's currently a loop around all authentication methods that outputs a single line depending on the authentication mode. This is not necessary as `HttpBasicAuth`, `ApiKeyAuth` and `OAuth` are always present in `../models/models`. Always emitting a single line that references all of these will prevent this error.

### in: cookie

```
components:
    securitySchemes:
        JWTAuthentication:
            type: apiKey
            in: cookie
            name: jwt
```

Will emit code that adds the API key to the query string, as `in: cookie` is not handled at all. This patch adds a handler that sets the `Cookie` header if this field is present.

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/typescript-node/models.mustache [hunk 1]
```diff
@@ -191,6 +191,13 @@ export class ApiKeyAuth implements Authentication {
             (<any>requestOptions.qs)[this.paramName] = this.apiKey;
         } else if (this.location == "header" && requestOptions && requestOptions.headers) {
             requestOptions.headers[this.paramName] = this.apiKey;
+        } else if (this.location == 'cookie' && requestOptions && requestOptions.headers) {
+            if (requestOptions.headers['Cookie']) {
+                requestOptions.headers['Cookie'] += '; ' + this.paramName + '=' + this.apiKey;
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-node/models.mustache]: Use encodeURIComponent() to encode the value

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache [hunk 2]
```diff
@@ -9,17 +9,7 @@ import { {{classname}} } from '../{{filename}}';
...
-import { HttpBasicAuth } from '../model/models';
-{{/isBasic}}
-{{#isApiKey}}
-import { ApiKeyAuth } from '../model/models';
-{{/isApiKey}}
-{{#isOAuth}}
-import { OAuth } from '../model/models';
-{{/isOAuth}}
-{{/authMethods}}
+import { HttpBasicAuth, ApiKeyAuth, OAuth } from '../model/models';
```
[the remark this answers is not in the document]
Comment (janjongboom) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: @macjohnny There are already places where unused imports are created in this exporter. E.g.: 

```yml
    schemas:
        GenericApiResponse:
            type: object
            required:
                - success
            properties:
                success:
                    type: boolean
                    description: Whether the operation succeeded
                error:
                    type: string
                    description: Optional error description (set if 'success' was false)
        GetJWTTokenResponse:
            allOf:
                - $ref: '#/components/schemas/GenericApiResponse'
                - type: object
                  properties:
                    token:
                        type: string
                        description: JWT token, to be used to log in in the future through JWTAuthentication
```

Yields in `getJWTTokenResponse.ts`:

```js
import { GenericApiResponse } from './genericApiResponse';
import { GetJWTTokenResponseAllOf } from './getJWTTokenResponseAllOf';

export class GetJWTTokenResponse {
```

Where `GenericApiResponse` and `GetJWTTokenResponseAllOf` are both imported, but not used in the file.

[the remark this answers is not in the document]
Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: ok, so then it hasn't been an issue so far

File: modules/openapi-generator/src/main/resources/typescript-node/models.mustache (hunk 1 above)
Comment (janjongboom) [reply modules/openapi-generator/src/main/resources/typescript-node/models.mustache]: Fixed and pushed update (rebased).

**Comment by macjohnny** [comment]
@ janjongboom please re-generate the samples, see https://circleci.com/gh/OpenAPITools/openapi-generator/9024#tests/containers/2

**Comment by janjongboom** [comment]
@macjohnny Sorry, didn't generate the openapi3 samples. Have pushed the updated files.

**Comment by janjongboom** [comment]
@macjohnny Pushed updated samples.

**Comment by macjohnny** [comment]
note: please run the https://github.com/OpenAPITools/openapi-generator/blob/master/bin/typescript-fetch-petstore-all.sh script, not the openapi3 version.

**Comment by janjongboom** [comment]
@macjohnny As far as I can see I ran both the `./bin` and the `./bin/openapi3` scripts... But I see you did a new push, so not sure if you require action from me at this point.

**Comment by macjohnny** [comment]
let's see what the CI says, but I think no more actions are required

**Comment by janjongboom** [comment]
Great, thanks for the quick review @macjohnny

**Comment by wing328** [comment]
@janjongboom thanks for the PR, which has been included in the v4.1.3 release: https://twitter.com/oas_generator/status/1180123829626003456

**Pull request merged**